src/api.ts

```diff
@@ -0,0 +1,559 @@
+  Request extends ApiRequest<any, any, any, any>,
+  IntegrationResponse,
+  MethodResponse
+> {
+  /**
+   * Function that maps an API request to an integration request and calls an
+   * integration. This will be compiled to a VTL request mapping template and
+   * an API GW integration.
+   *
+   * At present the function body must be a single statement calling an integration
+   * with an object literal argument. E.g
+   *
+   * ```ts
+   *  (req) => fn({ id: req.body.id });
+   * ```
+   *
+   * The supported syntax will be expanded in the future.
+   */
+  request: RequestTransformerFunction<Request, IntegrationResponse>;
+  /**
+   * Function that maps an integration response to a 200 method response. This
+   * is the happy path and is modeled explicitly so that the return type of the
+   * integration can be inferred. This will be compiled to a VTL template.
+   *
+   * At present the function body must be a single statement returning an object
+   * literal. The supported syntax will be expanded in the future.
+   */
+  response: ResponseTransformerFunction<IntegrationResponse, MethodResponse>;
+  /**
+   * Map of status codes to a function defining the  response to return. This is used
+   * to configure the failure path method responses, for e.g. when an integration fails.
+   */
+  errors: { [statusCode: number]: () => any };
+}
+
+/**
+ * An AWS API Gateway integration lets you integrate an API with an AWS service
+ * supported by Functionless. The request is transformed via VTL and sent to the
+ * service via API call, and the response is transformed via VTL and returned in
+ * the response.
+ *
+ * Only `application/json` is supported.
+ *
+ * TODO: provide example usage after api is stabilized
+ *
+ * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-api-integration-types.html
+ */
+export class AwsApiIntegration<
+  Props extends AwsApiIntegrationProps<any, any, any>
+> extends BaseApiIntegration {
+  private readonly request: FunctionDecl;
+  private readonly response: FunctionDecl;
+  private readonly errors: { [statusCode: number]: FunctionDecl };
+
+  public constructor(props: Props) {
+    super();
+    this.request = validateFunctionDecl(props.request);
+    this.response = validateFunctionDecl(props.response);
+    this.errors = Object.fromEntries(
+      Object.entries(props.errors).map(([k, v]) => [k, validateFunctionDecl(v)])
+    );
+  }
+
+  public addMethod(
+    httpMethod: HttpMethod,
+    resource: aws_apigateway.Resource
+  ): void {
+    const [requestTemplate, integration] = toVTL(this.request, "request");
+    const [responseTemplate] = toVTL(this.response, "response");
+
+    const errorResponses: aws_apigateway.IntegrationResponse[] = Object.entries(
+      this.errors
+    ).map(([statusCode, fn]) => {
+      const [template] = toVTL(fn, "response");
+      return {
+        statusCode: statusCode,
+        selectionPattern: `^${statusCode}$`,
```

Comment:
reusing `statusCode` for the `selectionPattern` isn't the best because it forces you to spell out every expected error instead of a catch all for all 5XX for example. Maybe something like this will work

```
errors: {
  "5\d\d": {
    statusCode: 500,
    fn: () => {...}],
  }
  ...
}
```